specs-derive/src/impl_saveload.rs

```diff
@@ -0,0 +1,487 @@
+                    .collect();
+                let field_ids_2 = field_ids.clone();
+                let tmp = field_ids.clone();
+
+                big_match_ser = quote!{
+                    #big_match_ser
+                    #name::#ident( #( ref #field_ids ),* ) => #saveload_name::#ident( #( ::specs::saveload::IntoSerialize::into(#field_ids_2, &mut ids)? ),* ),
+                };
+
+                let field_ids = tmp;
+                let field_ids_2 = field_ids.clone();
+
+                big_match_de = quote!{
+                    #big_match_de
+                    #saveload_name::#ident( #( #field_ids ),* ) => #name::#ident( #( ::specs::saveload::FromDeserialize::from(#field_ids_2, &mut ids)? ),* ),
+                };
+            }
+            Fields::Unit => {
+                big_match_ser = quote! {
+                    #big_match_ser
+                    #name::#ident => #saveload_name::#ident,
+                };
+
+                big_match_de = quote! {
+                    #big_match_de
+                    #saveload_name::#ident => #name::#ident,
+                };
+            }
+        }
+    }
+
+    let ser = quote!{
+        Ok(match *self {
+            #big_match_ser
+        })
+    };
+
+    let de = quote!{
+        Ok(match data {
+            #big_match_de
+        })
+    };
+
+    SaveloadDerive {
+        type_def: enum_def,
+        ser,
+        de,
+        saveload_name,
+        saveload_generics: saveload_generics.clone(), // dumb borrow checker, don't feel like fixing because it's probably NBD
+    }
+}
+
+// Edits the bounds of the input type to ensure all fields have `bound`.
+fn add_bound(generics: &mut Generics, bound: TypeParamBound) {
+    use syn::GenericParam;
+    for param in &mut generics.params {
+        if let GenericParam::Type(type_param) = param {
+            if type_param.bounds.iter().any(|b| bound == *b) {
+                continue;
+            }
+            type_param.bounds.push(bound.clone());
+        }
+    }
+}
+
+// Replaces the type with its corresponding `Data` type.
+fn replace_entity_type(ty: &mut Type) {
+    match ty {
+        Type::Array(ty) => replace_entity_type(&mut *ty.elem),
+        Type::Tuple(ty) => {
+            for ty in ty.elems.iter_mut() {
+                replace_entity_type(&mut *ty);
+            }
+        }
+        Type::Paren(ty) => replace_entity_type(&mut *ty.elem),
+        Type::Path(ty) => {
+            let ty_tok = ty.clone();
+            *ty = parse_quote!(<#ty_tok as ::specs::saveload::IntoSerialize<MA>>::Data);
+        }
+        Type::Group(ty) => replace_entity_type(&mut *ty.elem),
+
+        Type::TraitObject(_) => {}
+        Type::ImplTrait(_) => {}
+        Type::Slice(_) => {
+            panic!("Slices are unsupported, use owned types like Vecs or Arrays instead")
+        }
+        Type::Reference(_) => panic!("References are unsupported"),
+        Type::Ptr(_) => panic!("Raw pointer types are unsupported"),
```

Comment:
I didn't want to deal with reference/pointer graphs within serialized components (**especially** with cycles).

The biggest problem is it's not exactly clear what to even **do** with types that hold references while deserializing, since you'd need to somehow deserialize the contents of the reference and then pass that handle back to the callsite as well.

The funny thing is, we have a pointer that solves all these problems for us -- Entity, which is just a generational weak pointer. So I see no reason to try and solve these issues when we can tell the user "either write your own `impl` to explain what you mean by "(de)serialize this pointer" or split this into multiple components and use an `Entity` ID lookup".

Reply:
Oh right. I don't know what I was thinking there.